src/notifications.ts

```diff
@@ -0,0 +1,66 @@
+import { ChannelType, ThreadAutoArchiveDuration } from 'discord.js';
+import { getChannelFromSettings } from '@/util';
+import { NotificationThread } from '@/models/notificationThreads';
+import type { AllowedThreadTypeForTextChannel, BaseGuildTextChannel, Guild, GuildTextThreadCreateOptions, MessageCreateOptions, MessagePayload, ThreadChannel, User } from 'discord.js';
+
+export async function notifyUser(guild: Guild, user: User, message: string | MessageCreateOptions | MessagePayload): Promise<void> {
+	try {
+		await user.send(message);
+	} catch {
+		try {
+			await notifyUserInChannel(guild, user, message);
+		} catch (e) {
+			console.log(`Failed to notify user: ${user.displayName} (${user.id})`, e);
+		}
+	}
+}
+
+async function notifyUserInChannel(guild: Guild, user: User, message: string | MessageCreateOptions | MessagePayload): Promise<void> {
+	const notificationsChannel = await getChannelFromSettings(guild, 'channels.notifications');
+	if (!notificationsChannel || notificationsChannel.type !== ChannelType.GuildText) {
+		throw 'Could not find notifications channel';
+	}
+
+	let thread: ThreadChannel | null = null;
+	const cachedThreadMetadata = await NotificationThread.findOne({ where: { user_id: user.id } });
+	if (cachedThreadMetadata) {
+		try {
+			// * Need to wrap this in a `try` because even though `fetch` here says it can return `null` it
+			// * throws an exception if no channel is found for an ID
+			const cachedThread = await notificationsChannel.threads.fetch(cachedThreadMetadata.thread_id);
+			if (cachedThread) {
+				if (cachedThread.archived) {
+					await cachedThread.setArchived(false);
+					await cachedThreadMetadata.update({ updated: new Date() });
+				} 
+				thread = cachedThread;
+			} 
+		} catch (e) {
+			await cachedThreadMetadata.destroy();
+		}
+	}
+
+	if (!thread) {
+		thread = await createNotificationThread(notificationsChannel, user);
+	}
+
+	await thread.send(message);
+	await thread.members.add(user);
```

Comment:
My only issue with this is that it does not enforce a mention for notification threads. In the test server, the configuration I think works well is having the notifications channel not be assignable in onboarding, so it only shows up as "suggested" in the channel list to users for which it is relevant. Having a mention in the thread ensures that it shows up there. Users with their DMs open should never need to see the notifications channel because there would be no reason for it to be suggested.

![image](https://github.com/user-attachments/assets/c77289dd-acc0-4a1e-9deb-52f12777b6fa)

My original thought for this was that `notifyUser` could take an embed instead of a complete message, and it could set the content to nothing if it is a DM or a mention if it is in a notification thread. Another solution could be simply appending a mention to the beginning of the message and not including a mention in the call to `notifyUser`.

Reply:
We could optionally add an embed with a mention in it? Should be simple to do if I narrow the type to `string | MessageCreateOptions` because for `string` we can create a message from that and the embed, and for `MessageCreateOptions` we can append the embed?

Reply:
Unfortunately a mention in an embed is not a "real" mention and doesn't create a red "ping."

Reply:
My vote would be to just prepend a mention at the beginning of the message then. That way this still allows any message to be passed

Reply:
Setting the message to `<@user>\n${message}` sounds fine to me.